x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/DataFrameDataExtractorFactory.java

```diff
@@ -57,35 +59,38 @@
     private final String analyticsId;
     private final String index;
     private final ExtractedFields extractedFields;
+    private final QueryBuilder query;
 
-    private DataFrameDataExtractorFactory(Client client, String analyticsId, String index, ExtractedFields extractedFields) {
+    private DataFrameDataExtractorFactory(Client client, String analyticsId, String index, ExtractedFields extractedFields,
+                                          QueryBuilder query) {
         this.client = Objects.requireNonNull(client);
         this.analyticsId = Objects.requireNonNull(analyticsId);
         this.index = Objects.requireNonNull(index);
         this.extractedFields = Objects.requireNonNull(extractedFields);
+        this.query = query;
     }
 
     public DataFrameDataExtractor newExtractor(boolean includeSource) {
         DataFrameDataExtractorContext context = new DataFrameDataExtractorContext(
                 analyticsId,
                 extractedFields,
                 Arrays.asList(index),
-                QueryBuilders.matchAllQuery(),
+                query == null ? QueryBuilders.matchAllQuery() : query,
                 1000,
                 Collections.emptyMap(),
                 includeSource
             );
         return new DataFrameDataExtractor(client, context);
     }
 
-    public static void create(Client client, Map<String, String> headers, String analyticsId, String index,
+    public static void create(Client client, Map<String, String> headers, DataFrameAnalyticsConfig config,
```

Comment:
I updated this create method to just do what we expect, look at the source and create for the destination. 

If we would rather pass `analyticsId, index, query` separately, I can change it. I just found it confusing reading the code that we validate against the source (which never actually extracts anything), but then the task references the destination index.

Reply:
Unfortunately we cannot do this. In addition to what I mentioned above, there is another problem. Once we have reindexed `source` into the `dest` index, that becomes an immutable snapshot. The user might still be indexing stuff in the `source` index that we don't want to fetch.

Reply:
I did not think of that :(. I will adjust